### ISSUE TITLE
[RFC]: add `@stdlib/assert/is-complex-string`

### ISSUE THREAD

**Issue by marsian83** [stdlib-js/stdlib#1370]
### Description

This RFC proposes adding package `@stdlib/assert/is-complex-string`.
 
The package should be similar in structure to [`@stdlib/complex/reviver-float32`](https://github.com/stdlib-js/stdlib/tree/64931c5fdb4abc3a339b619658d3b5acc089e556/lib/node_modules/%40stdlib/complex/reviver-float32).

The goal is to assert if an input string is of a complex number format string`(a + ib)` or `(a+bi)`
where, a and b are numbers 

```js
 function isComplexString( str ) => boolean
```

Eg : isComplexString( "32+12i" ) returns true
Eg : isComplexString( "32+i12" ) returns true
Eg : isComplexString( null ) returns true
Eg : isComplexString( "" ) returns false
Eg : isComplexString( "abcd+12i" ) returns false
Eg : isComplexString( Infinity ) returns false
 
> Package: `@stdlib/assert/is-complex-string` Alias: `isComplexString`

### Related Issues

Can help with issues #1332 and #1333

### Questions

This is my first issue so I'd like to know if the way I have proposed the feature is correct

### Checklist

- [X] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [X] Searched for existing issues and pull requests.
- [X] The issue name begins with `RFC:`.

**Comment by marsian83** [comment]
I would like to implement this myself.
as a supporting assertion to #1332

**Comment by marsian83** [comment]
ok, so I took the initiative to start implementing this and I figured there would be 2 ways to approach this

one, using a regex 
```js
/^[-+]?\d*\.?\d+(?:[eE][-+]?\d+)?[-+]\d*\.?\d*(?:[eE][-+]?\d+)?i$/
```
This regex is quite popular as I have seen it everywhere on the internet while trying to look for a regex

secondly, I could implement a function to manually parse the string
something like this

```js

function isComplexString( str ) {
	if ( !isString( str ) ) {
		return false;
	}
	
    let re = '';
    let im = '';
    let hasIota = false;
    let i = 0;

    while (str[i] === ' ') {
        i++;
    }

    while (i < str.length && (str[i] >= '0' && str[i] <= '9' || str[i] === '.'
			|| str[i] === '+' || str[i] === '-')) {
        if (str[i] === 'i') {
            return false; // 'i' must not be in the real part
        }
        re += str[i++];
    }

    if (re.length === 0) {
        return false; // No real part
    }

    if (str[i] === '+' || str[i] === '-') {
        im += str[i++];
    }

    while (i < str.length && (str[i] >= '0' && str[i] <= '9' || str[i] === '.'
			|| str[i] === 'i')) {
        if (str[i] === 'i') {
            if (hasIota) {
                return false; // Duplicate 'i'
            }
            hasIota = true;
        }
        im += str[i++];
    }

    if (!hasIota) {
        return false; // No 'i' in the imaginary part
    }

    while (i < str.length) {
        if (str[i++] !== ' ') {
            return false; // non whitespace character after end of the string
        }
    }

    return true;
}


```

**Comment by marsian83** [comment]
which one of these approaches would be preferable according to the standards of other implementations in the codebas?
@Planeshifter  @kgryte

**Comment by kgryte** [comment]
You're right to identify that we need to handle exponential notation (see your regex).

```text
1 + 2i
1
2i
-2i
2i
1.523 - 4.234i
-1.5 + 4.2i
1.0e10 -2.0e-10i
1.0E10 + 4.1e+10i
1+2i
1-0i
0+1i
-0.0 + 2.0i
NaN + NaNi
Infinity-Infinityi
```

These should all be valid. Notice that we should also handle NaNs and infinities.

**Comment by kgryte** [comment]
My initial sense is that I suggest implementing a manual tokenizer/parser, as a regexp is more likely to be fiddly and given the variety of values, signs, etc, likely to become inscrutable.

**Comment by kgryte** [comment]
@marsian83 Perhaps before actually implementing this function and inlining the implementation, you should propose implementing `@stdlib/complex/base/parse-string`, which implements the tokenizer/parser and, if the string is valid, returns a complex-like object (e.g., `{ 're': 3.0, 'im': 5.0 }`) or, if the string is not valid, returns `null`.

Then in this package, the implementation is simply

```javascript
function isComplexString( value ) {
	return (
		isString( value ) &&
		parse( value ) !== null
	);
}
```

**Comment by marsian83** [comment]
> @marsian83 Perhaps before actually implementing this function and inlining the implementation, you should propose implementing `@stdlib/complex/base/parse-string`, which implements the tokenizer/parser and, if the string is valid, returns a complex-like object (e.g., `{ 're': 3.0, 'im': 5.0 }`) or, if the string is not valid, returns `null`.
> 
> Then in this package, the implementation is simply
> 
> ```js
> function isComplexString( value ) {
> 	return (
> 		isString( value ) &&
> 		parse( value ) !== null
> 	);
> }
> ```

Hi @kgryte , I have implemented the parse function but when trying to lint I have been facing this issue since my first issue

please Can you help me resolve this 
<img width="855" alt="image" src="https://github.com/stdlib-js/stdlib/assets/114365550/be323120-c5b2-4632-be04-17d7cffa9313">

**Comment by Codesmith28** [comment]
Hey there!
Can you assign me this issue?

**Comment by Planeshifter** [comment]
@marsian83 Please run `make init`, which will initialize the missing ESLint plugin.

**Comment by marsian83** [comment]
> @marsian83 Please run `make init`, which will initialize the missing ESLint plugin.

<img width="852" alt="image" src="https://github.com/stdlib-js/stdlib/assets/114365550/de2f464d-1a04-4a60-84d7-a9f385615bd6">

Hi, now I am just getting different errors

**Comment by Snehil-Shah** [comment]
> > @marsian83 Please run `make init`, which will initialize the missing ESLint plugin.
> 
> <img alt="image" width="852" src="https://private-user-images.githubusercontent.com/114365550/307598469-de2f464d-1a04-4a60-84d7-a9f385615bd6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg4NzE2ODgsIm5iZiI6MTcwODg3MTM4OCwicGF0aCI6Ii8xMTQzNjU1NTAvMzA3NTk4NDY5LWRlMmY0NjRkLTFhMDQtNGE2MC04NGQ3LWE5ZjM4NTYxNWJkNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyNVQxNDI5NDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04OTc4ZWYzNmVjZDBkMDI0NjgyZGZkODkxZDdhZGVlYmM2ZmU2MTBkOTc0MzZlNTVmZGJmOTFjZWVhOTg0MDVjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.rDfuDqzMvTfeDbIf5n7y4nNkxrBrcxaqby3Dxkf5p7E">
> Hi, now I am just getting different errors

I was getting these too. Try running `make install-node-modules` (after restoring the `package.json` that you might have lost). In my case, I was running it on WSL and setting up was really slow. If you're on windows, I suggest spinning up a Github Codespace instead of developing locally, it takes like 2 minutes to set up with no problems or errors.

**Comment by marsian83** [comment]
> @marsian83 Perhaps before actually implementing this function and inlining the implementation, you should propose implementing `@stdlib/complex/base/parse-string`, which implements the tokenizer/parser and, if the string is valid, returns a complex-like object (e.g., `{ 're': 3.0, 'im': 5.0 }`) or, if the string is not valid, returns `null`.
> 
> Then in this package, the implementation is simply
> 
> ```js
> function isComplexString( value ) {
> 	return (
> 		isString( value ) &&
> 		parse( value ) !== null
> 	);
> }
> ```

Hi @kgryte , I have made a PR with a possible implementation for this.

**Comment by marsian83** [comment]
Now that the `@stdlib/complex/base/parse` has been implemented 
This can now be implemented as well. I will get to working on this.
